server/handlers.go

```diff
@@ -475,6 +479,94 @@ func (s *Server) handleConnectorCallback(w http.ResponseWriter, r *http.Request)
 	http.Redirect(w, r, redirectURL, http.StatusSeeOther)
 }
 
+// handleSamlCallback handles a saml callback response with support for the IdP initiated flow.
+// if the RelayState is not one of our auth requests, we check to see if it is a valid redirectURI
+// for one of our clients. then, if the SAMLResponse is from a registered connector, we create a
+// new auth request in the database so we can contuine with our regular callback flow.
+func (s *Server) handleSamlCallback(r *http.Request) (string, string, int, error) {
+	relayState := r.PostFormValue("RelayState")
+	if relayState == "" {
+		return "", "", http.StatusBadRequest, fmt.Errorf("user session error")
+	}
+	// if our relaySate is a valid authId, this is not IdP initiated
+	if _, err := s.storage.GetAuthRequest(relayState); err == nil {
+		return relayState, relayState, http.StatusOK, nil
+	} else if err != storage.ErrNotFound {
+		s.logger.Errorf("Failed to get auth request: %v", err)
+		return "", "", http.StatusInternalServerError, fmt.Errorf("database error")
+	}
+	// find the client the RelayState is trying to send us to, stop when we match a redirectURI
+	// TODO: is it valid for differrent clients to have the same redirectURI? ¯\_(ツ)_/¯
+	clients, err := s.storage.ListClients()
+	if err != nil {
+		s.logger.Errorf("Failed to list clients: %v", err)
+		return "", "", http.StatusInternalServerError, fmt.Errorf("database error")
+	}
+	var client *storage.Client
+	for _, c := range clients {
+		for _, u := range c.RedirectURIs {
+			if relayState == u {
+				client = &c
+				break
+			}
+		}
+	}
+	if client == nil {
+		s.logger.Errorf("Cannot find client with redirectURI: %s", relayState)
+		return "", "", http.StatusInternalServerError, fmt.Errorf("bad saml response")
+	}
+	// find the issuer attribute in the SAMLResponse, use it to find the associated connector-id.
+	ssoIssuer, err := saml.GetSAMLIssuer(r.PostFormValue("SAMLResponse"))
+	if err != nil {
+		return "", "", http.StatusBadRequest, fmt.Errorf("bad saml response")
+	}
+	connectors, err := s.storage.ListConnectors()
+	if err != nil {
+		s.logger.Errorf("Failed to list connectors: %v", err)
+		return "", "", http.StatusInternalServerError, fmt.Errorf("database error")
+	}
+	var connID string
+	for _, c := range connectors {
+		if c.Type != "saml" {
+			continue
+		}
+		var cfg saml.Config
+		if err := json.Unmarshal(c.Config, &cfg); err != nil {
+			s.logger.Errorf("Cannot parse saml config for connector %s: %s", c.ID, err)
+			return "", "", http.StatusPreconditionFailed, fmt.Errorf("config error")
```

Comment:
yeeeah agreed that this shouldn't ever really happened, but wanted to handle the error regardless. since it's config related i figured it made sense as a precondition failure